### PR TITLE
Improve CI pinelines

### PR DESCRIPTION
**Pull request by danhphan**
This PR improves CI pipelines to address the issue #5 

It is a work-in-progress since I need to test several different configurations of github actions on Github hosted servers.

### CODE REVIEW

**Comment by danhphan** [comment]
The current commit has reduced the testing CI time from 1h15m to around 40m.

I am looking for further improvement on this, including more balanced test parts and bump to higher python versions (and remove python 3.7).

**Comment by danhphan** [comment]
The CI pipelines for tests now take around 26 mins. cc @michaelosthege @ricardoV94 @twiecki 

Let me know if it needs any further changes. Thanks!

PS: Current CI only has Ubuntu. There should be other CI tests for Windows and MacOs as well.

**Review by michaelosthege — COMMENTED** [review]
Wow!

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -67,27 +67,30 @@ jobs:
...
-          - python-version: "3.7"
+          - python-version: "3.10"
             fast-compile: 1
-            float32: 0
-            install-numba: 1
-            part: "tests"
+          - python-version: "3.10"
+            float32: 1
+          - python-version: "3.10"
+            part: "tests/tensor/test_math.py"
```
[the remark this answers is not in the document]
Comment (danhphan) [reply .github/workflows/test.yml:93]: Hi yes, the `exclude` will exclude this options: fast_compile == 1 && python-version == "3.10" && (float32 == 0 OR 1)

**Comment by danhphan** [comment]
The idea is somethings like that:

The `install-numba` is 1, and `part` options (how tests are separated) should be the same for all elements in the `strategy.matrix`.

So we have three variables left: `python-version`, `fast-compile`, and `float32`

In the previous test.yaml files, we run tests on these combinations:

- {python-version: 3.7, fast-compile: 0, float32: 0}
- {python-version: 3.7, fast-compile: 1, float32: 1} // using `include`
- {python-version: 3.7, fast-compile: 1, float32: 0} // using `include`
- {python-version: 3.9, fast-compile: 0, float32: 0}

In the PR, the three first combinations for **python 3.7** is updated to **(A)**:
- {python-version: 3.7, fast-compile: [0, 1], float32: [0, 1]}
- Exclude this case {python-version: 3.7, fast-compile: 0, float32: 1}

And the last combination for **python 3.9** is updated to **(B)**:
- {python-version: 3.9, fast-compile: [0, 1], float32: [0, 1]}
- Exclude this case {python-version: 3.9, fast-compile: 1}
- Exclude this case {python-version: 3.9, float32: 1}

Combine **(A)** and **(B)**, we will have:
- {python-version: [3.7, 3.9], fast-compile: [0, 1], float32: [0, 1]}
- Exclude this case {python-version: 3.7, fast-compile: 0, float32: 1}
- Exclude this case {python-version: 3.9, fast-compile: 1}
- Exclude this case {python-version: 3.9, float32: 1}

It seems that using `exclude` is clearer than `include`, and I do not want to repeat the `part` as it content is quite long with several lines of codes.

After that, changing python 3.7 => 3.8, and 3.9 => 3.10.

Hope that makes sense.

**Review by michaelosthege — APPROVED** [review]

**Review by Armavica — APPROVED** [review]

**Pull request merged**